README.md
# Lightning

Lightning is an Elixir framework for making simple REST API endpoints based on Plug

[![Hex version badge](https://img.shields.io/hexpm/v/lightning.svg)](https://hex.pm/packages/lightning)
[![Docs](https://img.shields.io/badge/docs-lightning-blue.svg)](https://hexdocs.pm/lightning/Lightning.html)



# How to get started:

Create a new file (App.ex)


```elixir
defmodule App do 
   use Lightning.HTTP
   import Lightning

 #Create a new route endpoint
 #Route: GET "/helloworld/"
 def route("GET", ["helloworld"], conn) do

    #Set additional response information (based on Plug responses)
    conn 
    |> put_resp_header("Hello", "World")
    |> put_resp_content_type("text/html")
    |> put_resp_cookie("abc", "def")
    |> put_status(200)

    #Send a text response with a statuscode of 200:
    text(conn, 200, "Hello world")
   end
end
```
# Running the server
Start up a server using the iex command:
```elixir
    iex -S mix
    iex> {:ok, _} = Lightning.start(5000, App, :dev)
    
```

Navigating to localhost:5000/helloworld will output text response:<br>
```Hello world```


# Example Code

![code example](https://i.imgur.com/DYQREls.png)


lightning/CHANGELOG.md
# Changelog
(dd-mm-yyyy)


## v0.1.8 (14-12-2018)
	#FIXED: Adding response data now work correctly
	* v0.1.8 API CHANGES
```elixir
	#Omitted the use of the res keyword for response operations
		  	FROM: def route("GET", ["text"], conn, res) do
			TO:   def route("GET", ["text"], conn) do

			FROM: text(conn, res, 200, "Hello World")
			TO:	  text(conn, 200, "Hello World")
````
```elixir
	#NEW: global response function:
			FROM: text(conn, 200, "Hello World")
				  json(conn, 200, %{"hello" => "world"})
				  eex(conn, 200, Path.expand("./lib/templates/show_user.eex"),[user: user])

			TO:	response(conn, 200, "Hello World")
				response(conn, 200, %{"hello" => "world"})
				response(conn, 200, Path.expand("./lib/templates/show_user.eex"),[user: user])

			* json/3, text/3, eex/4, can still be used normally alongside the new response function
````


## v0.1.7 (13-12-2018)
	* Added annotations BETA
	* v0.1.7 API CHANGES
```elixir
	#Changed reponse API
		FROM: send_text(conn, res, 200, "Hello World")
		TO:	  text(conn, 200, "Hello World")

		FROM: send_json(conn, res, 200, %{"Hello" => "World"})
		TO:	  json(conn, 200, %{"Hello" => "World"})

		FROM: send_eex(conn, res, 200, Path.expand("./lib/templates/show_user.eex"), [greeting: "Hello"])
		TO:	  eex(conn, 200, Path.expand("./lib/templates/show_user.eex"), [greeting: "Hello"])


	#All code does not have to be prefixed by the Lightning.* 
````


## v0.1.6 (12-12-2018)
	* v0.1.6 API CHANGES
```elixir
	#NEW: parse_key function:
		#EXAMPLE:
			name = parse_key(conn, :name)

	#Changed architecture of core Modules
	#Using the Lightning library now requires:
		FROM: use Lightning

		TO:	  use Lightning.HTTP
			  import Lightning

	#All code does not have to be prefixed by the Lightning.*
````

## v0.1.5 (11-12-2018)
	* v0.1.5 API CHANGES
```elixir
	#NEW: pattern matching functionality for parse_body function:
		FROM: name = Lightning.parse_body(conn).params["name"]
		TO:   [name, age] = Lightning.parse_body(conn)
````

## v0.1.4 (11-12-2018)
	* Fixed critical bug that prevented the server from booting up


## v0.1.3 (9-12-2018)
	* v0.1.3 API CHANGES
```elixir
	#Created abstraction (Lightning.start) for server startup
		FROM: {:ok, _} = Plug.Adapters.Cowboy.http App, []
		TO:   {:ok, _} = Lightning.start(5000, App, :dev)
````
			

## v0.1.2 (9-12-2018)
	* v0.1.2 API CHANGES
	* Updated docs to reflect name changes
	* Added TODO file
```elixir
	#Changed ambiguous naming of module Lightning.HTTP to Lightning
		FROM: use Lightning.HTTP<br>
		TO:   use Lightning
````
	* Updated docs to reflect name changes
	* Added TODO file


## v0.1.1 (7-12-2018)
	* Updated docs
	* fixed errors in docs
	* Added logo


## v0.1.0 (7-12-2018)
	  * First release


lightning/lib/HTTP/Lightning.ex
defmodule Lightning do
    require EEx

    @moduledoc """

    Lightning is a framework for making simple REST API endpoints based on Plug

    WARNING: API subject to change in future 0.x versions.<br>
    check changelog for breaking API changes: 
    <a href="https://github.com/CasperCX/Lightning/blob/master/lightning/CHANGELOG.md">CHANGELOG</a> 

    ## How to get started:

        Create a new file (App.ex)

        defmodule App do

        # Use the Lightning library
        use Lightning.HTTP
        import Lightning
   
        # New Route: GET "/helloworld/"
        def route("GET", ["helloworld"], conn) do

            # Set additional response information (based on Plug responses):
            conn 
            |> put_resp_header("Header", "Here")
            |> put_resp_content_type("application/json")
            |> put_resp_cookie("abc", "def")
            |> put_status(200)
            
            # Send an JSON response with a statuscode of 200:
            json(conn, 200, %{"hello" => "world"})
        end

        -----------------------------------------
        Start up a server using the iex command:
            iex -S mix
            iex> {:ok, _} = Lightning.start(5000, App, :dev)

        # Navigating to localhost:5000/helloworld will output JSON response:
        # {"hello":"world"}

    """

    # TODO
    # def start(port, modules, _environment \\ :dev) when is_list(modules) do
        # modules |> Enum.each(fn(mod) -> {:ok, _} = Plug.Cowboy.http(mod, [], port: port))
    # end


    @doc """
    Start the server.

    ## Start up the server

        EXAMPLE:

        Lightning.start(<port>, <module_name>, <environment>)

        iex -S mix
        iex> {:ok, _} = Lightning.start(5000, App, :dev)
        
        # Locate to localhost:5000/ in the browser

    """
    def start(port, module, _environment) do
        {:ok, _} = Plug.Cowboy.http(module, [], port: port)

        # TODO hotloading config
        # case environment do
        #     :dev -> IO.puts "start hotloading"
        # end
    end


  


    def response(conn, status, body) when is_bitstring(body) do
         conn |> Plug.Conn.send_resp(status, body)
    end


    def response(conn, status, body) when is_map(body) do
        resp_body = body |> Poison.encode!()

        conn |> Plug.Conn.send_resp(status, resp_body)
    end


    def response(conn, status, template, vars \\ []) do
        case vars do
        # Precompiled version
        [] -> conn |> Plug.Conn.send_resp(status, template)

        _  ->  page_contents = EEx.eval_file(template, vars) 
                conn |> Plug.Conn.send_resp(status, page_contents)
        end
    end




    @doc """
    Text Response.

    ## return a text response

        EXAMPLE:

        # Route: GET "/hello/:firstname/:lastname/"
         def route("GET", ["hello", firstname, lastname], conn) do

            # Set additional response information:
            conn 
            |> put_resp_header("Header", "Here")
            |> put_resp_content_type("text/html")

            # Send an text response with a statuscode of 200:
            text(conn, 200, "Hello from text response " <> firstname <> lastname)
        end

    """
    def text(conn, status, body) do
        conn |> Plug.Conn.send_resp(status, body)
        # conn |> res.send_resp(status, body)
    end



    @doc """
    JSON Response.

    ## return a JSON response

        EXAMPLE:

        # Route: GET "/json/"
        def route("GET", ["json"], conn) do

            # Set additional response information:
            conn 
            |> put_resp_header("Header", "Here")
            |> put_resp_content_type("application/json")

            # Send an JSON response with a statuscode of 200:
            json(conn, 200, %{"age" => 26, "name" => "My_name"})
        end

    """
    def json(conn, status, body) do
        resp_body = body |> Poison.encode!()

        conn |> Plug.Conn.send_resp(status, resp_body)
    end



    @doc """
    EEx file response.

    ## return an EEx response

        EXAMPLE:

        # Route: GET "/user/:user_id/"
        def route("GET", ["user", user_id], conn) do

            # Set additional response information
            conn 
            |> put_resp_header("Header", "Here")
            |> put_resp_content_type("text/html")

            # Example using Ecto, returning different responses based on condition:
            case App.Repo.get(User, user_id) do
                nil 
                    -> text(conn, 404, "User with ID " <> user_id <> " not found")
                user 
                    -> eex(conn, 200, Path.expand("./lib/templates/show_user.eex"), [user: user, user_id: user_id])
            end
        end


        ---------------------------------------------
        Alternatively you can pre-compile EEx files (faster).
        # place at top of App.ex file:

        defmodule App do
            require EEx
            EEx.function_from_file(:def, :template_show_user, Path.expand("./lib/templates/show_user.eex"), [:user, :user_id])

        # and now use its pre-compiled version passing in the user and user_id as arguments:
        eex(conn, 200, template_show_user(user, user_id)) #
        
    """
    def eex(conn, status, template, vars \\ []) do
        case vars do
            # Precompiled version
            [] -> conn |> Plug.Conn.send_resp(status, template)

            _  ->  page_contents = EEx.eval_file(template, vars) 
                    conn |> Plug.Conn.send_resp(status, page_contents)

        end   
    end



    @doc """
    Parse the incoming request body.

    ## Parse incoming request body

        EXAMPLE:

        # Route: POST "/parse?name=casper&age=26&sex=male
        def route("POST", ["parse"], conn) do

            # patternmatch the request body and get the values of first two keys and ignoring last
            [name, age, _] = parse_body(conn)

            # Return a text response with status code 200
            text(conn, 200, "Hello, name: " <> name <> " age: " <> age)
        end

    """
     def parse_body(conn, opts \\ []) do
        opts = Keyword.put_new(opts, :parsers, [Plug.Parsers.URLENCODED, Plug.Parsers.MULTIPART])

        # Get the values from the param keys and return as a list
        Plug.Parsers.call(conn, Plug.Parsers.init(opts)).query_params |> Map.values
    end



     @doc """
    Parse the incoming request key.

    ## Parse incoming key

        EXAMPLE:

        # Route: POST "/parse?name=casper
        def route("POST", ["parse"], conn) do

            # Request the value of key: name
            name = parse_key(conn, :name)

            # Return a text response with status code 200
            text(conn, 200, "Hello, name: " <> name)
        end

    """
    def parse_key(conn, key, opts \\ []) do
        opts = Keyword.put_new(opts, :parsers, [Plug.Parsers.URLENCODED, Plug.Parsers.MULTIPART])

        # Get the values from the param keys and return as a list
        Plug.Parsers.call(conn, Plug.Parsers.init(opts)).params[Atom.to_string(key)]
    end

    def redirect(conn, path_name) do
        # TODO
        # Look up if path_name is a registered route
        # lightning.Annotations.annotation

        # GOTO route url defined in annotation
        # conn |> Plug.Conn.

        case path_name do
                ["favicon.ico"] ->
                    %{conn | path_info: path_name}
                _ ->
                     %{conn | path_info: path_name}
            end
        end

         # def redirect_index(%Plug.Conn{path_info: path} = conn, _opts) do
            #     case path do
            #     [] ->
            #         %{conn | path_info: ["static", "index.html"]}

            #     ["favicon.ico"] ->
            #         %{conn | path_info: ["static", "favicon.ico"]}

            #     _ ->
            #         conn
            #     end
            # end
   


    #TODO
    def parse_json(conn, status, body) do
        IO.puts "parsing JSON"

        conn |> Plug.Conn.send_resp(status, body)
    end

end





    #    CODE SNIPPETS FOR RESEARCH

    #    use Plug.Router
        # plug Plug.Static,
        #     at: "/public",
        #     from: :my_app,
        #     only_matching: ~w(favicon)

        # defmodule HttpServer.Router do
            # use Plug.Router

            # plug(Plug.Logger)
            # plug(:redirect_index)
            # plug(:match)
            # plug(:dispatch)

            # forward("/static", to: HttpServer.StaticResources)

            # get "/sse" do
            #     # some other stuff...
            #     conn
            # end

            # match _ do
            #     send_resp(conn, 404, "not found")
            # end

            # def redirect_index(%Plug.Conn{path_info: path} = conn, _opts) do
            #     case path do
            #     [] ->
            #         %{conn | path_info: ["static", "index.html"]}

            #     ["favicon.ico"] ->
            #         %{conn | path_info: ["static", "favicon.ico"]}

            #     _ ->
            #         conn
            #     end
            # end

    # res = Plug.Conn
    # case conn.path_info do
    #         ["favicon.ico"] ->
    #             # %{conn | path_info: ["static", "favicon.ico"]}
    #             conn

    #         [] ->
    #             route(conn.method, conn.path_info, conn, res)

    #         # _ ->
    #         #     conn
    #     end
    
    
    # # case conn.path_info do
    # #     ["favicon.ico"] -> IO.puts("Favicon called")




    # use Plug.Builder
    # # use Plug.Router
    # plug(:redirect_index)

    # match _ do
    #     send_resp(conn, 404, "not found")
    # end

    # def redirect_index(conn, path_name \\ :path_name) do
    #     # case conn.path_info do
    #     # [] ->
    #     #     %{conn | path_info: [Atom.to_string(path_name)]}

    #     # ["favicon.ico"] ->
    #     #     %{conn | path_info: ["static", "favicon.ico"]}

    #     # _ ->
    #     #     conn
    #     # end

    #         %{conn | path_info: ["ya"]}
    #         # %{conn | path_info: [Atom.to_string(path_name)]}
    # end
    # def redirect(conn) do
        

    #     %{conn | path_info: ["ya"]}
    # end